### PR TITLE
struct equal limit

### PR DESCRIPTION
**Pull request by Ashuaidehao**
Close https://github.com/neo-project/neo-vm/issues/426

### CODE REVIEW

**Comment by superboyiii** [comment]
Need fix UT.

**Comment by roman-khimov** [comment]
Looks good, should fix the problem.

**Review by shargon — APPROVED** [review]

**Comment by dusmart** [comment]
> Looks good, should fix the problem.

Although the problem is fixed, too many limits are introduced. Here is my concern.

If we think of the `struct` as a value, we'd better make sure the nested reference of struct won't appear. No two struct fields can point to the same object.

In this case, we construct a struct like this by using `dup` and `pack` as well as `convert`.

A                     (new struct)
A'                    (dup)
B -> A             (pack
B -> A'             pack + convert)
B'                     (dup)
C -> B              (pack
C -> B'             pack + convert)

As shown in this case, the value turns too big because of the ref. We should never let a **value** contain two refs to the same object.

**Comment by vang1ong7ang** [comment]
Yeah

If `Struct` is a value type, it could be better to never introduce its ref into the stack.

If `Struct` is a ref type, no additional `Equal` function is needed, just compare their ref.

**Comment by roman-khimov** [comment]
> Currently Struct is more like something between value and ref, which sometimes misleading the dev imo.

I can agree with that, but changing it right now is just too dangerous, all compilers assume this behaviour.

**Comment by superboyiii** [comment]
I agree with roman. We can apply it on v3.0.0 first and then we have more time to make a better solution later (might change a lot).

**Comment by superboyiii** [comment]
@erikzhang What do you think?

**Comment by ProDog** [comment]
Seems to be working, execute `560112c501fe0160589d604a12c0db415824f7509d4a102aec4597`, throw `System.InvalidOperationException: Too many struct items to compare`.

File: src/neo-vm/Types/Struct.cs [hunk 1]
```diff
@@ -68,6 +69,11 @@ public override StackItem ConvertTo(StackItemType type)
         }
 
         public override bool Equals(StackItem? other)
+        {
+            return InternalEquals(other, ExecutionEngineLimits.Default.MaxStackSize);
```
Comment (erikzhang) [inline src/neo-vm/Types/Struct.cs:73]: What if the VM uses different values then `ExecutionEngineLimits.Default`?

Comment (Ashuaidehao) [reply src/neo-vm/Types/Struct.cs:73]: I think this limit should be a const value not a config value.

**Review by erikzhang — APPROVED** [review]

**Comment by erikzhang** [comment]
Please test it again.

**Comment by ProDog** [comment]
> Please test it again.

Tested OK.

**Pull request merged**